Add ability to specify assert() calls in embed.fnc

khwilliamson opened the pull request:
This allows the removal of the G flag, and the removal of three customized PERL_ARGS_ASSERT macros.  One of these showed the brittleness of the old scheme, as after it was written, code was added to improve the generic versions of these macros, but no one realized to update the custom one.  This can no longer happen with this pull request

This commit also improves the error checking of embed.fnc entries; adds comments and white space

* This set of changes includes a perldelta entry

khwilliamson committed:
regen/embed.pl: Skip unnecessary code

If an argument is a literal string, it's not going to be anything else,
so convert an elsif to an else.

khwilliamson committed:
regen/embed.pl: White-space only

Indent newly formed block from the previous commit

khwilliamson committed:
regen/embed.pl: Improve error checking

This commit verifies that only one of the mutually exclusive argument
modifiers, like NN, is present, and that NZ isn't used for a pointer
argument.

khwilliamson committed:
regen/HeaderParser.pm: Add comments; white-space

khwilliamson committed:
regen/HeaderParser.pm: Put all non-word flags at end

This generalizes the previous code to make sure all non-word flags are
sorted after the word ones.  Previously it checked for just '#'

All but '#' not being in column 1 is just for aesthetics.  If '#' were
in column 1 it would be a preprocessor directive that likely would fail
to compile.

khwilliamson committed:
embed.fnc: Add ability to customize ARGS_ASSERT macros

This enhances embed.fnc entries to contain arbitrary assert() calls,
which will be added to the corresponding generated ARGS_ASSERT macros.

The next commit will use it.

khwilliamson committed:
Remove customized ARGS_ASSERT macros

These existed because they exceeded the rudimentary capabilities of the
generated ARGS_ASSERT macros.  But the previous commit generalized that
capability, so they can change to use that.

This moves the three extant customized macros to use the new facility,
removing the need for overriding things

The customized PERL_ARGS_ASSERT_ISA_LOOKUP was out-of-date, missing an
assertion that got added, without the macro getting updated.  This new
scheme will prevent that from happening in the future.

khwilliamson committed:
embed.fnc: Remove spurious G flag from entry

This prevented an empty PERL_ARGS_ASSERT macro from being generated
here, but those are optional to use, so there is no advantage to
suppressing it.

bram-perl commented on proto.h:
```diff
@@ -9450,12 +9452,16 @@ S_utf16_textfilter(pTHX_ int idx, SV *sv, int maxlen);
-S_sv_derived_from_svpvn(pTHX_ SV *sv, SV *namesv, const char *name, const STRLEN len, U32 flags);
+S_sv_derived_from_svpvn(pTHX_ SV *sv, SV *namesv, const char *name, const STRLEN len, U32 flags                              );
+# define PERL_ARGS_ASSERT_SV_DERIVED_FROM_SVPVN \
+        assert(namesv || name)
```
**Commit Remove customized ARGS_ASSERT macros**

This needs updating;

The old macro:
```
#define PERL_ARGS_ASSERT_SV_DERIVED_FROM_SVPVN \
    assert(sv); \
    assert(namesv || name)
```

In the new macro the `assert(sv)` is missing. I'm **assuming** it was marked as NULLOK in embed.fnc to avoid the macro being generated. (Before it would always generate the macro for NN arguments)

khwilliamson replied:
You are right; thank you

bram-perl commented on embed.fnc:
```diff
@@ -443,13 +443,6 @@
 :        flag even on a format function is if the format would generate error:
 :        format string argument is not a string type
 :
-:   'G'  Suppress empty PERL_ARGS_ASSERT_foo macro.  Normally such a macro is
```
**Commit embed.fnc: Remove no longer used G flag**

It's still being used in regen/embed.pl:
```
        my $args_assert_line = ( $flags !~ /[Gm]/ );
```
I suppse the `G` should be removed from the regex.

khwilliamson replied:
Yes, and actually in two regexes

khwilliamson committed:
embed.fnc: Remove no longer used G flag

The purpose of this flag has been removed in the past few commits

khwilliamson committed:
Move assert for av_new_alloc() to embed.fnc

You can't allocate a zero length array using this function; this uses
the new assert() functionality in embed.fnc introduced in the previous
few commits to cause the assertion to be placed in
PERL_ARGS_ASSERT_AV_NEW_ALLOC rather than be in the function.

There is no particular need to do this; this is mostly a demonstration
of how it can be done.

leonerd reviewed (commented):
These changes look good, but it'd be useful to squash the commits down a bit. We don't need all 11 distinct ones.

Maybe one commit to add the new assert ability, one to swap the existing uses of the `G` flag and custom macros, and one final one to remove support for that `G` flag.

bram-perl commented on regen/HeaderParser.pm:
```diff
@@ -749,12 +749,16 @@ sub tidy_embed_fnc_entry {
     my ($flags, $ret, $name, @args)= split /\s*\|\s*/, $line;
 
-    # Sort and remove duplicate flags
+    # Sort and remove duplicate flags.  Alpha flags are sorted first
```
**Commit regen/HeaderParser.pm: All non-word flags come at end**

Can you rephrase the commit title/message slightly?
Reading **only** the commit title it's not clear what to expect; it sounds more like a statement then about making a change. (I **think** that when seeing only the title I would've expected a doc-fix or something but not a code change).
Maybe something like: `Put non-word flags at the end`? (Or "Put all ...")

(When completing the commit message with: `If applied, this commit will ...`, i.e: `If applied, this commit will all non-word flags come at end` then it's a sentence that just sounds awkward and doesn't make sense.)

khwilliamson replied:
Agreed; done, I think.

bram-perl commented on utf8.c:
```diff
@@ -125,6 +125,7 @@ U8 *
 Perl_uvoffuni_to_utf8_flags(pTHX_ U8 *d, UV uv, const UV flags)
 {
     PERL_ARGS_ASSERT_UVOFFUNI_TO_UTF8_FLAGS;
+    // XXX macroize
```
**Commit utf8.c-comment**

Should probably not be included in this PR?

khwilliamson replied:
done

bram-perl commented on regen/embed.pl:
```diff
@@ -247,19 +247,30 @@ sub generate_proto_h {
                     my $nn =      ( $arg =~ s/\s*\bNN\b\s+// );
                     my $nz =      ( $arg =~ s/\s*\bNZ\b\s+// );
                     my $nullok =  ( $arg =~ s/\s*\bNULLOK\b\s+// );
                     my $nocheck = ( $arg =~ s/\s*\bNOCHECK\b\s+// );
```
**Commit regen/embed.pl: Improve error checking**

Unrelated to this commit but this code is actually buggy and it will cause your new code not to detect (all) errors.
(The problem already existed so it's not new.)

Starting point:
- `$arg = "NULLOK NN SV *msg_sv";`
- after: `$nn = ...`  it is changed into: `"NULLOKSV *msg_sv"`
- after `$nullok = ...` it remains as: `"NULLOKSV *msg_sv"`
- result: `$nn=1`, `$nullok=undef`
- (end result: incorrect/broken macro)

The issue is that `$nn = ...`  replaces both the leading and trailing white space which caused it to become `NULLOKSV` instead of `NULLOK SV` (note the space).

A **possible** fix (untested) could be do replace it with a single space (i.e. `$arg =~ s/\s*\bNN\b\s+/ /`) and then later remove leading/duplicated/trailing whitespace.

khwilliamson replied:
Thanks; nice catch

jkeenan commented:
> > These changes look good, but it'd be useful to squash the commits down a bit. We don't need all 11 distinct ones.
> 

I would appreciate some squashing along the lines suggested by @leonerd because it would reduce slightly the number of commits that need to be traversed in a bisection used to diagnose bugs in the future.

> What I would personally do is merge it with a merge-commit so that it's clear it's one block of related changes. But that does not appear to be something that is typically done in blead. 

```
$ git log --format=fuller --date=short --since 2024-08-07 --reverse | grep -c '^commit '
2048
$ git log --format=fuller --date=short --since 2024-08-07 --reverse | grep -c '^Merge: '
14
```
Fewer than 1% of all commits in the last 365 days -- but almost all of them were the culmination of projects long in development and were performed by major contributors.  Even within such merges, however, some squashing can be beneficial.

bram-perl commented on regen/HeaderParser.pm:
```diff
@@ -737,46 +737,63 @@ sub lines { $_[0]->{lines} }
+    # Sort and remove duplicate flags
     my %flag_seen;
-    $flags= join "", grep !$flag_seen{$_}++, sort split //, $flags;
+    $flags = join "", grep !$flag_seen{$_}++, sort split //, $flags;
```
**Commit regen/HeaderParser.pm: Add comments; white-space**

Just an observation and something to think about; (not an objection.)

The rest of the file uses dmq's style which is: `$x= ...`. You changed it to `$x = ...` which makes it now less consistent with the rest of the file.

(Personally I use and prefer `$x = ...` over `$x= ...` but having consistency in the same file can be important too.)

khwilliamson replied:
I hadn't realized this was a style; I thought it was just a typo, because it violates perlstyle that says that most operators should be separated from their operands by a space.  It doesn't mention which ones are the exceptions, but unaries  like `++`  are what I think they are.  Many people don't separate unary `!` from its operand, but I find that hard to read.  They likely will too as they get older.  I'm ok with `5|7` not having spaces.  And I follow the common style of not using spaces to do arithmetic on array subscripts.  `a[i+1]`

I believe it is a mistake to add code that violates our style rules to conform to surrounding code.  What our policy is is leave such things alone except when you  touch such a line for some valid reason.

bram-perl commented on regen/HeaderParser.pm:
```diff
@@ -749,12 +749,16 @@ sub tidy_embed_fnc_entry {
-    }
+    $flags = join "", grep !$flag_seen{$_}++,
+                      sort {
+                             my $a_is_word = $a =~ /\w/;
+                             my $b_is_word = $b =~ /\w/;
+                             return $a cmp $b if $a_is_word == $b_is_word;
+                             return -1 if $a_is_word;
+                             return  1;
```
**Commit regen/HeaderParser.pm: All non-word flags come at end**

Consider adding a test for this in t/porting/header_parser.t.

khwilliamson replied:
I don't believe this is worth a test.  It's just a nicety.  Everything builds fine without this except if '#' gets placed in column 1.  And that likely would generate a syntax error as an improper preprocessor directive.  If it somehow broke, and nobody noticed, it would be fine.

khwilliamson commented:
I've done a few bisections recently.  We're at 40K commits when the bisect kicks in.  We don't need an extra step until we reach 65536 commits.  That's more than half again as many as we've ever done.  That means that trying to conserve commits won't help bisecting.

Anything we can do to save human time helps us.  Bite-sized commits are helpful in this regard.

khwilliamson committed:
regen/embed.pl: Don't meld words to get wrong results

@Bram-perl found this example of how it could be broken before this
commit:

    initial  : $arg is "NULLOK NN SV *msg_sv
    after s//: $arg is "NULLOKSV *msg_sv

bram-perl commented on regen/embed.pl:
```diff
@@ -274,6 +274,25 @@ sub generate_proto_h {
+                    # pointer.
+                    if ($args_assert_line && $arg =~ /\*/) {
+                        if ($nn + $nullok == 0) {
+                            warn "$func: $arg needs NN or NULLOK\n";
+                            ++$unflagged_pointers;
```
**Commit regen/embed.pl: Improve error checking**

This commit needs rechecking;.

Same commit in the previous push: https://github.com/Perl/perl5/commit/a2819421e5edfd099747d8695619220284ea7fb4 
--> it removed the block on line 259-265:
```diff
-                  if (   $args_assert_line
-                        && $arg =~ /\*/
-                        && $arg !~ /\b(NN|NULLOK)\b/ )
-                    {
-                        warn "$func: $arg needs NN or NULLOK\n";
-                        ++$unflagged_pointers;
-                    }
-
```

In this commit it's kept.. 

I believe the block on lines 259-265 should be removed because the same check is now done here.

khwilliamson replied:
This was a rebase -i error.  I'm not sure if it was me or rere.  Both possibilities are scary (that I missed it)

bram-perl commented on regen/embed.pl:
```diff
@@ -121,7 +121,8 @@ sub generate_proto_h {
         my $inner_ind= $ind ? "  " : " ";
 
-        my ($flags,$retval,$plain_func,$args) = @{$embed}{qw(flags return_type name args)};
+        my ($flags, $retval, $plain_func, $args, $assertions ) =
+                        @{$embed}{qw(flags return_type name args assertions)};
```
**Commit embed.fnc: Add ability to customize ARGS_ASSERT macros**

Just a side note: this is a value slice of a hash ref, it can also be written using postderef as:
```
my ($flags, $retval, $plain_func, $args, $assertions ) = $embed->@{qw(flags return_type name args assertions)};
```

(Not asking for a change; just offering it as hint, feel free to ignore it.)

khwilliamson replied:
These lines at the beginning explain why not:

```
require 5.004;  # keep this compatible, an old perl is all we may have before
                # we build the new one
```

khwilliamson replied:
(That's not to say that that dictum has been obeyed in all commits; and maybe the number should be changed to more recent.)

bram-perl replied:
In that case you should update the line:
```perl
        push @asserts, $assertions->@* if $assertions;
```

which you added since that too is using post-deref.

khwilliamson replied:
Yeah; I imagine there are others that have crept in.

khwilliamson committed:
perldelta for embed.fnc assert() statements

khwilliamson commented:
I have now added a perldelta entry

khwilliamson commented:
My rules for what needs to be separate commits, is "Would it ever be better to keep this change if we needed to revert the rest?"  And sometimes for indentation changes,  "Does this change make another change's difference listing easier enough to grok to be worth it?"

khwilliamson commented:
Closed via e6095bad29472df96aa0896de9304bcf1ba4f17c
